vignettes/dp-histogram.Rmd

```diff
@@ -0,0 +1,247 @@
+
+```{r}
+character_histogram <- dpHistogram(var.type='character', variable="Employee.Status", n=469, epsilon=0.5,
+                                   bins=c("Detailee","Employee","Employee (part-time)"))
+character_histogram$release(SocrataWhiteHouseEmployeeSalaries)
+print(character_histogram$result)
+```
+
+If you would like NA (empty or unknown) values to be imputed (or replaced) by a specific subset of input bins, you can specify the imputation bins with `impute.bins`. You may want to do this if you know the underlying data is skewed, meaning an unkown value is more likely to be of a certain value as opposed to others. Specifying `impute.bins` is optional, and if it is not specified the NA values ill be imputed using a uniform distribution on all imput bins (i.e. NA values will be replaced with any known value with equal probability).
+
+```{r}
+character_histogram <- dpHistogram(var.type='character', variable="Employee.Status", n=469, epsilon=0.5,
+                                   bins=c("Detailee","Employee","Employee (part-time)"), impute.bins=c("Detailee","Employee"))
+character_histogram$release(SocrataWhiteHouseEmployeeSalaries)
+print(character_histogram$result)
+```
+
+&nbsp;
+
+What is the Stability Mechanism and How is it used?
+------------
+
+There may be cases where you do not have prior knowledge about the data, and you are making a histogram to learn more about the data. Alternatively, there may be cases where you do not want certain features of the data to be released or revealed with the histogram. In these cases, the stability mechanism will be used to add noise to the histogram to create a differentially private release instead of the general Laplace mechanism. In general, the stability mechanism is one which takes advantage of “stable” functions, i.e. ones where the function output is constant in some neighborhood around the input database. In this library, the stability mechanism is implemented to be used specifically for the histogram statistic. For a histogram generated by the Stability mechanism, empty buckets will be removed and any buckets with a count below an accuracy threshold will also be removed. Removing these buckets based on low counts is what creates the added guarantee of privacy.
+
+If you enter a list of bins for any type of variable, the Lapalce mechanism will be used. Similarly, the Laplace mechanism will always be used for logical variables.
+
+If you do not enter bins for categorical variables, the stability mechanism will be used. The stability mechanism wil determine appropriate bins for the histogram, and will remove any bins with too low of a count, which may breach privacy. The example below is the same as the character histogram with the bins entered above, except the bins are removed and a delta value is entered. Notice how the result is different because the stability mechanism is used:
```

Comment:
Maybe explain a bit how this is a privacy breach? I.e. the problem here is that if you release empty bins you give away information about the other categories in the database, which is information you previously did not have.

Reply:
how is it that you release info about other bins? I thought the problem was that you release information about the empty bins?